jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PushLayerStep.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.builder.steps;
+
+import com.google.cloud.tools.jib.api.RegistryException;
+import com.google.cloud.tools.jib.blob.BlobDescriptor;
+import com.google.cloud.tools.jib.builder.ProgressEventDispatcher;
+import com.google.cloud.tools.jib.builder.TimerEventDispatcher;
+import com.google.cloud.tools.jib.cache.CachedLayer;
+import com.google.cloud.tools.jib.configuration.BuildConfiguration;
+import com.google.cloud.tools.jib.http.Authorization;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+class PushLayerStep implements Callable<BlobDescriptor> {
+
+  private static final String DESCRIPTION = "Preparing application layer pushers";
+
+  private final BuildConfiguration buildConfiguration;
+  private final ProgressEventDispatcher.Factory progressEventDispatcherFactory;
+
+  private final Authorization pushAuthorization;
+  private final Future<CachedLayerAndName> cachedLayerAndName;
+
+  static ImmutableList<PushLayerStep> makeList(
+      BuildConfiguration buildConfiguration,
+      ProgressEventDispatcher.Factory progressEventDispatcherFactory,
+      Authorization pushAuthorization,
+      List<Future<CachedLayerAndName>> cachedLayers) {
```

Comment:
Note to reviewers: this is the only exception where a step gets a "future" value. This is to allow pushing layers even if Jib is still downloading other layers. In any case, the class is greatly simplified, and no concurrency-related scheduling is being done inside a step. The pattern to create sub-tasks is same as pulling base image layers and building application layers.